### PR TITLE
feat[cartesian]: Specify compiler optimization level and flags via env variables

### PR DESCRIPTION
**Pull request by stubbiali**
## Description

This PR allows downstream applications to specify compiler optimization level and flags on a stencil basis.

## Requirements

- [ ] All fixes and/or new features come with corresponding tests.
- [ ] Important design decisions have been documented in the approriate ADR inside the [docs/development/ADRs/](docs/development/ADRs/Index.md) folder.

If this PR contains code authored by new contributors please make sure:

- [ ] All the authors are covered by a valid contributor assignment agreement provided to ETH Zurich and signed by the employer if needed.
- [ ] The PR contains an updated version of the `AUTHORS.md` file adding the names of all the new contributors.

### CODE REVIEW

**Review by havogt — CHANGES_REQUESTED** [review]
Please add a test that illustrates using the feature. Not sure if removing `debug_mode` is a good idea.

**Comment by havogt** [comment]
Adding @egparedes and @FlorianDeconinck for more opinions.

**Review by FlorianDeconinck — APPROVED** [review]
LGTM

File: src/gt4py/cartesian/backend/pyext_builder.py [hunk 1]
```diff
@@ -112,9 +113,9 @@ def get_gt_pyext_build_opts(
             "--compiler-options",
             "-fPIC",
         ]
-    extra_link_args = gt_config.build_settings["extra_link_args"]
+    extra_link_args = copy.deepcopy(gt_config.build_settings["extra_link_args"])
 
-    mode_flags = ["-O0", "-ggdb"] if debug_mode else ["-O3", "-DNDEBUG"]
+    mode_flags = [f"-O{opt_level}", "-DNDEBUG", *extra_opt_flags.split()]
```
Comment (stubbiali) [inline src/gt4py/cartesian/backend/pyext_builder.py:118]: @havogt On (my) mac, the flag `-DNDEBUG` is required to make compilation works with any optimization level, so I hardcoded it into `mode_flags` for the sake of convenience. If you prefer, we could be more flexible: 
```suggestion
    mode_flags = [f"-O{opt_level}"]
    if extra_opt_flags != "":
        mode_flags += [*extra_opt_flags.split()]
    elif opt_level == "0":
        mode_flags.append("-ggdb")
    else:
        mode_flags.append("-DNDEBUG")
```

File: src/gt4py/cartesian/backend/pyext_builder.py [hunk 2]
```diff
@@ -47,7 +47,8 @@ def get_cuda_compute_capability():
 
 def get_gt_pyext_build_opts(
     *,
-    debug_mode: bool = False,
+    opt_level: Literal["0", "1", "2", "3", "s"] = "3",
```
Comment (havogt) [inline src/gt4py/cartesian/backend/pyext_builder.py:50]: Semantically I would do the following:
```suggestion
    opt_level: Literal["0", "1", "2", "3", "s"] = "3" if not debug_mode else "0",
```
I.e. keep the old behavior for `debug_mode`, but if you use additionally the opt_level, then overwrite the default opt_level.

File: src/gt4py/cartesian/backend/pyext_builder.py [hunk 3]
```diff
@@ -112,9 +114,12 @@ def get_gt_pyext_build_opts(
             "--compiler-options",
             "-fPIC",
         ]
-    extra_link_args = gt_config.build_settings["extra_link_args"]
+    extra_link_args = copy.deepcopy(gt_config.build_settings["extra_link_args"])
+
+    mode_flags = (
+        ["-O0", "-ggdb"] if debug_mode else [f"-O{opt_level}", "-DNDEBUG", *extra_opt_flags.split()]
```
Comment (havogt) [inline src/gt4py/cartesian/backend/pyext_builder.py:120]: this fixes my comment above

**Review by havogt — APPROVED** [review]

**Comment by havogt** [comment]
cscs-ci run

**Pull request merged**